Bài 1 Thực hiện lại các hướng dẫn từ trang 89 đến 90.


In [1]:
from sklearn.datasets import load_boston
X, y = load_boston(return_X_y=True)
print(X.shape)
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(
	X, y, test_size=0.2, random_state=1)
from sklearn import svm
from sklearn.model_selection import GridSearchCV

tuned_parameters = [{'kernel' : ['rbf'], 'C' : [1,10,100]},
                   {'kernel' : ['linear'], 'C' : [1,10,100]}]
svm_reg = GridSearchCV (svm.SVR(), tuned_parameters, scoring = 'neg_mean_absolute_error', cv=2
                       )
svm_reg.fit(X_train, y_train)
svm_reg.best_params_

/home/toandokhanh/.local/lib/python3.10/site-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function load_boston is deprecated; `load_boston` is deprecated in 1.0 and will be removed in 1.2.

    The Boston housing prices dataset has an ethical problem. You can refer to
    the documentation of this function for further details.

    The scikit-learn maintainers therefore strongly discourage the use of this
    dataset unless the purpose of the code is to study and educate about
    ethical issues in data science and machine learning.

    In this special case, you can fetch the dataset from the original
    source::

        import pandas as pd
        import numpy as np

        data_url = "http://lib.stat.cmu.edu/datasets/boston"
        raw_df = pd.read_csv(data_url, sep="\s+", skiprows=22, header=None)
        data = np.hstack([raw_df.values[::2, :], raw_df.values[1::2, :2]])
        target = raw_df.values[1::2, 2]

    Alternative datasets include the California housing

(506, 13)


{'C': 1, 'kernel': 'linear'}

In [2]:
from sklearn import svm
svm_regr = svm.SVR(kernel='linear', C=1)
svm_regr.fit(X_train, y_train)

from sklearn.metrics import mean_squared_error, mean_absolute_error
y_train_pred = svm_regr.predict(X_train)
print('Train: Mean squared error = ', mean_squared_error(y_train,y_train_pred))
print('Train: Mean absolute error = ', mean_absolute_error(y_train,y_train_pred))

y_test_pred = svm_regr.predict(X_test)
print('Test: Mean squared error = ', mean_squared_error(y_test,y_test_pred))
print('Test: Mean absolute error = ', mean_absolute_error(y_test,y_test_pred))


Train: Mean squared error =  25.00568757175539
Train: Mean absolute error =  3.025409980645767
Test: Mean squared error =  23.845815281597613
Test: Mean absolute error =  3.481070310170902


Câu 3: Trong phần code lựa chọn siêu tham số ở [11] hình 4.31 trang 90 với 2 siêu tham số là kernel và C, bạn hãy tự đề xuất những giá trị khác của bộ siêu tham số này (Vd: C=[50, 100], kernel = ‘poly’, ‘rbf’, ‘sigmoid’,....) để đạt được độ lỗi tốt hơn trong giáo trình (Hiện giá trị MAE bộ test trong giáo trình là 3.48). Cho biết giá trị bộ siêu tham số đạt kết quả tốt nhất là gì (vd: trong giáo trình là C=1, kernel=linear sẽ đạt kết quả tốt nhất).


In [7]:
from sklearn.datasets import load_boston
X, y = load_boston(return_X_y=True)
print(X.shape)
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(
	X, y, test_size=0.2, random_state=1)

from sklearn import svm
from sklearn.model_selection import GridSearchCV

tuned_parameters = [{'kernel': ['poly'], 'C': [50, 100]},
                   {'kernel': ['sigmoid'], 'C': [50, 100]},
                   {'kernel': ['rbf'], 'C': [50, 100]}]

svm_reg = GridSearchCV(
    	svm.SVR(), tuned_parameters,
    	scoring = 'neg_mean_absolute_error', cv=2
	)
svm_reg.fit(X_train, y_train)
svm_reg.best_params_

(506, 13)


/home/toandokhanh/.local/lib/python3.10/site-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function load_boston is deprecated; `load_boston` is deprecated in 1.0 and will be removed in 1.2.

    The Boston housing prices dataset has an ethical problem. You can refer to
    the documentation of this function for further details.

    The scikit-learn maintainers therefore strongly discourage the use of this
    dataset unless the purpose of the code is to study and educate about
    ethical issues in data science and machine learning.

    In this special case, you can fetch the dataset from the original
    source::

        import pandas as pd
        import numpy as np

        data_url = "http://lib.stat.cmu.edu/datasets/boston"
        raw_df = pd.read_csv(data_url, sep="\s+", skiprows=22, header=None)
        data = np.hstack([raw_df.values[::2, :], raw_df.values[1::2, :2]])
        target = raw_df.values[1::2, 2]

    Alternative datasets include the California housing

{'C': 100, 'kernel': 'poly'}

In [8]:
from sklearn import svm
svm_regr = svm.SVR(kernel='linear', C=2)
svm_regr.fit(X_train, y_train)

from sklearn.metrics import mean_squared_error, mean_absolute_error
y_train_pred = svm_regr.predict(X_train)
print('Train: Mean squared error = ',mean_squared_error(y_train, y_train_pred))
print('Train: Mean absolute error = ',mean_absolute_error(y_train, y_train_pred))

y_test_pred = svm_regr.predict(X_test)
print('Test: Mean squared error = ',mean_squared_error(y_test, y_test_pred))
print('Test: Mean absolute error = ',mean_absolute_error(y_test, y_test_pred))

Train: Mean squared error =  25.046030357013468
Train: Mean absolute error =  3.017728118194072
Test: Mean squared error =  23.470666040587954
Test: Mean absolute error =  3.4495178304792375


câu 4: Sử dụng khối [1] ở hình 4.13 trang 78 để load dữ liệu digits, kết hợp với [11] ở hình 4.31 để lựa chọn siêu tham số cho bài toán phân lớp 10 số ở dataset digits. Gợi ý: Dòng 8 khối lệnh 11, bạn sửa “svm.SVR()” (dành cho bài toán hồi quy-regression) thành “svm.SVC” (dành cho bài toán phân lớp - classification); dòng 9 sửa “scoring=neg…” thành “scoring=f1” (dựa vào độ đo f1-score để chọn siêu tham số tốt nhất). Chạy và cho biết bộ siêu tham số nào tốt nhất, chạy phân lớp với bộ siêu tham số vừa chọn (khối lệnh [12] hình 4.32). Dùng classification_report để thể hiện kết quả trên tập huấn luyện và tập kiểm tra.

In [13]:
from sklearn.datasets import load_digits
X, y = load_digits(return_X_y=True)
print(X.shape)
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(
	X, y, test_size=0.2, random_state=1)

(1797, 64)


In [14]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import  classification_report
from sklearn.model_selection import GridSearchCV

tuned_parameters = [{'penalty': ['none'], 'solver': ['newton-cg', 'lbfgs', 'liblinear']}]

log_reg = GridSearchCV(
    	LogisticRegression(), tuned_parameters,
    	scoring='f1', cv=2
	)
log_reg.fit(X_train, y_train)
log_reg.best_params_

/home/toandokhanh/.local/lib/python3.10/site-packages/sklearn/model_selection/_validation.py:776: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: 
Traceback (most recent call last):
  File "/home/toandokhanh/.local/lib/python3.10/site-packages/sklearn/model_selection/_validation.py", line 767, in _score
    scores = scorer(estimator, X_test, y_test)
  File "/home/toandokhanh/.local/lib/python3.10/site-packages/sklearn/metrics/_scorer.py", line 219, in __call__
    return self._score(
  File "/home/toandokhanh/.local/lib/python3.10/site-packages/sklearn/metrics/_scorer.py", line 267, in _score
    return self._sign * self._score_func(y_true, y_pred, **self._kwargs)
  File "/home/toandokhanh/.local/lib/python3.10/site-packages/sklearn/metrics/_classification.py", line 1136, in f1_score
    return fbeta_score(
  File "/home/toandokhanh/.local/lib/python3.10/site-packages/sklearn/metrics/_classification.py", line 1277, i

{'penalty': 'none', 'solver': 'newton-cg'}

In [15]:
from sklearn.linear_model import LogisticRegression

log_reg = LogisticRegression(penalty='none', solver='lbfgs')
log_reg.fit(X_train, y_train)

from sklearn.metrics import classification_report

y_train_pred = log_reg.predict(X_train)
print('Train: classification report = ')
print(classification_report(y_train, y_train_pred))
y_test_pred = log_reg.predict(X_test)
print('Test: classification report = ')
print(classification_report(y_test, y_test_pred))


Train: classification report = 
              precision    recall  f1-score   support

           0       1.00      1.00      1.00       135
           1       1.00      1.00      1.00       147
           2       1.00      1.00      1.00       141
           3       1.00      1.00      1.00       142
           4       1.00      1.00      1.00       143
           5       1.00      1.00      1.00       152
           6       1.00      1.00      1.00       144
           7       1.00      1.00      1.00       142
           8       1.00      1.00      1.00       145
           9       1.00      1.00      1.00       146

    accuracy                           1.00      1437
   macro avg       1.00      1.00      1.00      1437
weighted avg       1.00      1.00      1.00      1437

Test: classification report = 
              precision    recall  f1-score   support

           0       1.00      0.98      0.99        43
           1       0.97      0.91      0.94        35
           2   

Câu 5: Như câu trên nhưng thực hiện với thuật toán LogisticRegression. Gợi ý: Bạn thế chỗ svm.SVR thành LogisticRegression(), thế chỗ dòng 1: “from sklearn.linear_model import LogisticRegression”, thế độ đo như câu 9, phần tham số các bạn đọc ở đây để biết tham số của LogisticRegression (gợi ý: solver,...). Dùng classification_report để thể hiện kết quả trên tập huấn luyện và tập kiểm tra.


In [16]:
from sklearn.datasets import load_digits
X, y = load_digits(return_X_y=True)
print(X.shape)

(1797, 64)


In [17]:
from sklearn.metrics import  classification_report
from sklearn import svm
from sklearn.model_selection import GridSearchCV

tuned_parameters = [{'kernel': ['rbf'], 'C': [1, 10, 100]},
                   {'kernel': ['linear'], 'C': [1, 10, 100]}]

svm_reg = GridSearchCV(
    	svm.SVC(), tuned_parameters,
    	scoring = 'f1', cv=2
	)
svm_reg.fit(X_train, y_train)
svm_reg.best_params_

/home/toandokhanh/.local/lib/python3.10/site-packages/sklearn/model_selection/_validation.py:776: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: 
Traceback (most recent call last):
  File "/home/toandokhanh/.local/lib/python3.10/site-packages/sklearn/model_selection/_validation.py", line 767, in _score
    scores = scorer(estimator, X_test, y_test)
  File "/home/toandokhanh/.local/lib/python3.10/site-packages/sklearn/metrics/_scorer.py", line 219, in __call__
    return self._score(
  File "/home/toandokhanh/.local/lib/python3.10/site-packages/sklearn/metrics/_scorer.py", line 267, in _score
    return self._sign * self._score_func(y_true, y_pred, **self._kwargs)
  File "/home/toandokhanh/.local/lib/python3.10/site-packages/sklearn/metrics/_classification.py", line 1136, in f1_score
    return fbeta_score(
  File "/home/toandokhanh/.local/lib/python3.10/site-packages/sklearn/metrics/_classification.py", line 1277, i

/home/toandokhanh/.local/lib/python3.10/site-packages/sklearn/model_selection/_validation.py:776: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: 
Traceback (most recent call last):
  File "/home/toandokhanh/.local/lib/python3.10/site-packages/sklearn/model_selection/_validation.py", line 767, in _score
    scores = scorer(estimator, X_test, y_test)
  File "/home/toandokhanh/.local/lib/python3.10/site-packages/sklearn/metrics/_scorer.py", line 219, in __call__
    return self._score(
  File "/home/toandokhanh/.local/lib/python3.10/site-packages/sklearn/metrics/_scorer.py", line 267, in _score
    return self._sign * self._score_func(y_true, y_pred, **self._kwargs)
  File "/home/toandokhanh/.local/lib/python3.10/site-packages/sklearn/metrics/_classification.py", line 1136, in f1_score
    return fbeta_score(
  File "/home/toandokhanh/.local/lib/python3.10/site-packages/sklearn/metrics/_classification.py", line 1277, i

{'C': 1, 'kernel': 'rbf'}

In [18]:
from sklearn import svm

svm_regr = svm.SVC(kernel='rbf', C=100)
svm_regr.fit(X_train, y_train)

from sklearn.metrics import mean_squared_error, mean_absolute_error
y_train_pred = svm_regr.predict(X_train)
print('Train: Mean squared error = ')
print(classification_report(y_train, y_train_pred))
print('Train: Mean absolute error = ')
print(classification_report(y_train, y_train_pred))

y_test_pred = svm_regr.predict(X_test)
print('Test: Mean squared error = ')
print(classification_report(y_test, y_test_pred))
print('Test: Mean absolute error = ')
print(classification_report(y_test, y_test_pred))

Train: Mean squared error = 
              precision    recall  f1-score   support

           0       1.00      1.00      1.00       135
           1       1.00      1.00      1.00       147
           2       1.00      1.00      1.00       141
           3       1.00      1.00      1.00       142
           4       1.00      1.00      1.00       143
           5       1.00      1.00      1.00       152
           6       1.00      1.00      1.00       144
           7       1.00      1.00      1.00       142
           8       1.00      1.00      1.00       145
           9       1.00      1.00      1.00       146

    accuracy                           1.00      1437
   macro avg       1.00      1.00      1.00      1437
weighted avg       1.00      1.00      1.00      1437

Train: Mean absolute error = 
              precision    recall  f1-score   support

           0       1.00      1.00      1.00       135
           1       1.00      1.00      1.00       147
           2       